pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -20,148 +20,296 @@
+   * A = 100 OR A = 200 OR A = 300 OR A IN (200, 300, 400, 500, 600)
+   *
+   * We simply remove OR and return a FilterQueryTree with leaf operator
+   * -- if the number of values is 1, use EQUALITY
+   * -- else use IN
+   *
+   * If this scenario is not applicable, we look at both the set of operators
+   * separately
+   *
+   * First check if EQUALITY/IN set of child operators need to be rewritten.
+   * We already store this information in map for each column, so just check from there.
+   *
+   * Next check if non EQUALITY/IN operators need to be rewrriten. The only
+   * case where we there is a possibility is if the operator is AND since
+   * we don't know what is under AND
+   *
+   * Examples to demonstrate this:
+   *
+   * a IN (1) OR b IN (2) -- two IN operators and zero other operators
+   * both IN operators can be rewritten to EQUALITY
+   *
+   * a IN (100) OR b = 300 -- one IN and one EQUALITY operator,
+   * zero other operators, IN can be rewritten
+   *
+   * a = 100 OR b = 300 -- two equality and zero other operators
+   * nothing needs to be rewritten
+   *
+   * a = 1 OR b = 2 OR c > 20 -- two equality and one other operator
+   * nothing needs to be rewritten
 
-      // If we have at least one column to return
-      if (!columnToValues.isEmpty()) {
-        // We can eliminate the OR node if there is only one column with multiple values
-        if (columnToValues.size() == 1 && nonEqualityOperators.isEmpty()) {
-          Map.Entry<String, Set<String>> columnAndValues = columnToValues.entrySet().iterator().next();
+   * a IN (1) OR b IN (2) OR c > 20 -- two IN operators and one other operators
+   * both IN operators can be rewritten to EQUALITY and other operator need not
+   * be rewritten -- we simply append to child list of new root
+   *
+   * a IN (1) OR b IN (2) OR (c > 20 AND a > 200)
+   * two IN operators and AND other operator. both IN can be rewritten
+   * and for AND we can't decide unless we drill down so we recurse --
+   * the recursion will return the AND tree as is and we append it to
+   * the list of of children for OR
+   *
+   * a = 1 OR b = 2 OR (c > 20 AND a > 200)
+   * two EQUALITY operators and AND other operator.
+   * both EQUALITY need not be rewritten
+   * and for AND we can't decide unless we drill down so we recurse --
+   * the recursion will return the AND tree as is and we append it to
+   * the list of of children for OR
+   *
+   * a = 1 OR b = 2 OR (c IN (20) AND a > 200)
+   * two EQUALITY operators and AND other operator.
+   * both EQUALITY need not be rewritten
+   * and for AND we can't decide unless we drill down so we recurse --
+   * the recursion will return the AND tree with the inner IN rewritten
+   *
+   * @param filterQueryTree root of tree to optimize
+   * @return root of optimized tree
+   */
+  private FilterQueryTree optimizeTreeRootedAtOR(FilterQueryTree filterQueryTree) {
+    Map<String, Set<String>> columnToInEqPredicateValues = new HashMap<>();
```

Comment:
TreeMap might be better as numColumns should be very limited?
Also order will be deterministic

Reply:
Yes probably. However, since amortized O(1) v/s O(logN) might impact the runtime of this algorithm a little bit (and not sure how much if there is a pathological query with deep tree and several columns). So I'd say we can do that in a follow-up PR accompanied by a benchmark. We can also use TreeSet for the inner predicate values.

I have added a TODO